src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs

```diff
@@ -57,17 +57,26 @@ public async Task StartAsync(CancellationToken cancellationToken = default)
             await _hostLifetime.WaitForStartAsync(combinedCancellationToken).ConfigureAwait(false);
 
             combinedCancellationToken.ThrowIfCancellationRequested();
-            _hostedServices = Services.GetService<IEnumerable<IHostedService>>();
+            var hostedServices = Services.GetService<IEnumerable<IHostedService>>();
 
-            foreach (IHostedService hostedService in _hostedServices)
+            foreach (IHostedService hostedService in hostedServices)
             {
+                if (_applicationLifetime.ApplicationStopping.IsCancellationRequested)
```

Comment:
Thank you for this feedback 👍 The reasoning for this implementation is to get consistent behavior for calls to `IHostApplicationLifetime.StopApplication()`, despite the state of the host. I would argue that it is expected to get a graceful stop even though the call is made before the host is fully started.

However, I think that if the user provided `CancellationToken` is cancelled, it should throw, since this is expected behavior from other APIs that accept a `CancellationToken`.

So to sum it up: before starting each hosted service, check if the application is instructed to stop, if so, return from the Start method and let the host begin continue to stop. Next, check if the user has requested cancellation, and if so throw exception.

Looking forward hearing your thoughts on this!

Reply:
Did you see @davidfowl's comments here?

https://github.com/dotnet/runtime/pull/52517#discussion_r629436543

> It's weird to return from StartAsync without an error if we haven't successfully started.

Reply:
Maybe it's fine but I haven't thought enough about the implications. Is there anything that would be broken if the started event didn't fire but StartAsync returns successfully?